### ISSUE TITLE
[BUG] DR restored data is faulty in sles15-sp2

### ISSUE THREAD

**Issue by c3y1huang** [longhorn/longhorn#2503]
**Describe the bug**
The DR volume restored from the previous backup delta block looks weird and inconsistent with the original volume data. 

The middle of DR volume data looks weird (zero string?). Seems blocks from the last backup are correct but the data of the beginning of the previous backup or end of the last backup contains some empty bytes.

Verified this does not happen in Ubuntu.

![Screenshot-2021-04-16-11:05:51](https://user-images.githubusercontent.com/22583541/115012187-1a8a4980-9ee2-11eb-9fb3-a858baab218d.png)

**To Reproduce**
1.Longhorn cluster on AWS sles15-sp2.
2. Run `test_dr_volume_with_backup_block_deletion`.
```
Summary of the testcase steps:
1. Create backup-0 with 2 delta blocks for data.
2. Overwrite 1st block of data and create backup-1
3. Create DR volume from backup-1
4. Delete backup-1 from the original volume.
5. Overwrite 1st block of data and create backup-2
6. Verify DR volume and original volume data.
```

**Expected behavior**
Pass, DR volume data and original volume data should be the same.

**Log**
n/a

**Environment:**
 - Longhorn version: master (~1ae9b9b)
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s
   - Number of management node in the cluster: 1
   - Number of worker node in the cluster: 2
 - Node config
   - OS type and version: sles15-sp2
   - CPU per node: 4
   - Memory per node: 8
   - Disk type(e.g. SSD/NVMe): ssd
   - Network bandwidth between the nodes: 10G
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): AWS
 - Number of Longhorn volumes in the cluster: -

**Additional context**
https://github.com/longhorn/longhorn-tests/pull/592

**Comment by c3y1huang** [comment]
This is also happening in v1.1.0.

**Comment by c3y1huang** [comment]
`test_restore_inc` passed for both sles15-sp2 and ubuntu.

**Comment by c3y1huang** [comment]
> `test_restore_inc` passed for both sles15-sp2 and ubuntu. cc/ @shuo-wu

However, if I replace the backup1 content with random data, it will also fail `test_restore_inc`.

**Comment by innobead** [comment]
cc @khushboo-rancher @meldafrawi

**Comment by c3y1huang** [comment]
The file block number created during the DR 1st restore is almost twice the expected (8064 vs 4096).
```
&{volume-delta-backup-ea3168eca9194b00.img 2097152 420 {799041747 63754486743 0x18dd860} {66305 260071716 1 33188 0 0 0 0 2097152 4096 8064 {1618889943 755041250} {1618889943 799041747} {1618889943 799041747} [0 0 0]}}" pkg=backupstore

&{volume-delta-backup-8d05a16d362b4e85.img 16777216 420 {147957663 63754500480 0x18dc860} {66305 9861263 1 33188 0 0 0 0 16777216 4096 4096 {1618903680 99957418} {1618903680 147957663} {1618903680 147957663} [0 0 0]}}
```

The cause is coming from the [io.CopyN](https://github.com/longhorn/backupstore/blob/master/deltablock.go#L512). Could be due to the misalignment. Replacing with [Write]((f *File) Write() brought down the block number to expected.

**Comment by shuo-wu** [comment]
> The file block number written in sles during the 1st restore is almost twice the expected (8064 vs 4096).
> 
> ```
> &{volume-delta-backup-ea3168eca9194b00.img 2097152 420 {799041747 63754486743 0x18dd860} {66305 260071716 1 33188 0 0 0 0 2097152 4096 8064 {1618889943 755041250} {1618889943 799041747} {1618889943 799041747} [0 0 0]}}
> 
> &{volume-delta-backup-8d05a16d362b4e85.img 16777216 420 {147957663 63754500480 0x18dc860} {66305 9861263 1 33188 0 0 0 0 16777216 4096 4096 {1618903680 99957418} {1618903680 147957663} {1618903680 147957663} [0 0 0]}}
> ```
> 
> This is coming from the [io.CopyN](https://github.com/longhorn/backupstore/blob/master/deltablock.go#L512). Possibly due to the misalignment. Replacing with [Write](https://golang.org/pkg/os/#File.Write) brought down the block number to the expected.

The 1st restore? You mean the data of the DR volume is not right in the beginning? Is `r` of `io.CopyN(volDev, r, DEFAULT_BLOCK_SIZE)` correct in SLES?

BTW, what does the inserted part mean?

**Comment by c3y1huang** [comment]
> The 1st restore? You mean the data of the DR volume is not right in the beginning? 
The data is not right at the beginning and seems to be coming from the wrong block number.

> Is `r` of `io.CopyN(volDev, r, DEFAULT_BLOCK_SIZE)` correct in SLES?
The r is correct in SLES, I've compared the byte length and they are identical.
> 
> BTW, what does the inserted part mean?
Where did you see that...?

**Comment by shuo-wu** [comment]
Oh I mean "insert code", which is
```
&{volume-delta-backup-ea3168eca9194b00.img 2097152 420 {799041747 63754486743 0x18dd860} {66305 260071716 1 33188 0 0 0 0 2097152 4096 8064 {1618889943 755041250} {1618889943 799041747} {1618889943 799041747} [0 0 0]}}

&{volume-delta-backup-8d05a16d362b4e85.img 16777216 420 {147957663 63754500480 0x18dc860} {66305 9861263 1 33188 0 0 0 0 16777216 4096 4096 {1618903680 99957418} {1618903680 147957663} {1618903680 147957663} [0 0 0]}}
```

**Comment by c3y1huang** [comment]
> Oh I mean "insert code", which is
> 
> ```
> &{volume-delta-backup-ea3168eca9194b00.img 2097152 420 {799041747 63754486743 0x18dd860} {66305 260071716 1 33188 0 0 0 0 2097152 4096 8064 {1618889943 755041250} {1618889943 799041747} {1618889943 799041747} [0 0 0]}}
> 
> &{volume-delta-backup-8d05a16d362b4e85.img 16777216 420 {147957663 63754500480 0x18dc860} {66305 9861263 1 33188 0 0 0 0 16777216 4096 4096 {1618903680 99957418} {1618903680 147957663} {1618903680 147957663} [0 0 0]}}
> ```

They are from the FileInfo and the part relevant is from [Stat_t](https://golang.org/pkg/syscall/#Stat_t).

**Comment by c3y1huang** [comment]
This issue is found on XFS filesystem due to the speculative preallocation algorithm.

[The XFS speculative preallocation algorithm allocates extra blocks beyond end of file (EOF) to minimize file fragmentation during buffered write workloads.](https://xfs.org/index.php/XFS_FAQ#Q:_Why_do_files_on_XFS_use_more_data_blocks_than_expected.3F)

Taking a backup with 2Mib of data trying to restore to a 16Mib volume as an example.

The expected formation of the volume file would be
data(2Mib)+ zero bytes(16 - 2Mib)

However the actual formation of the volume file on xfs end up with
data(2Mib) + xfs post-EOF blocks(?Mib) + zero bytes(16 - 2 - ? Mib)

This is because extra blocks are allocated post-EOF.
```
             EOF
delta blocks  | preallocated blocks     |
--------------+-------------------------+
data blocks     extra allocated blocks  |
```

[In most cases, speculative preallocation is automatically reclaimed when a file is closed.](https://xfs.org/index.php/XFS_FAQ#Q:_What_is_speculative_preallocation.3F)
```
             EOF
delta blocks  |
--------------+
data blocks   |
```

But since we've truncated the file before the file is closed, it moves the EOF the caused the file unable to find and reclaim the speculative preallocation at post-EOF. And this resulted in the extra fragmented file blocks.
```
                                                           EOF
delta blocks  |  preallocated blocks     |  truncate block  |
--------------+--------------------------+------------------+
data blocks      extra allocated blocks     zero bytes      |
```

The fix for this is to truncate the file before backup data is restored. This ensures the EOF does not get relocated and the preallocated blocks get reclaimed at file close.

This is a concern with the current and prior versions when using backup to restore. User is likely to encounter and suffer file fragmentation when using XFS filesystem for backup restore with the current and prior version. If just backing up and not restore, it should not be a concern after the fix.

If the user is unable to wait until the next release to use backup restore, they should have the data restored to a new cluster on the ext4 file system.

**Comment by yasker** [comment]
Thanks for the great write-up @c3y1huang ! It's clear now.

The `preallocated blocks` is unlikely 0. So we will have different data stored in the restored file than what's backed up.

I am trying to access the potential impact:
1. If the user is using the volume as a raw block device without filesystem, then they might be in trouble. Though the majority of the user are using the volume with filesystem. 
2. If the user is using a filesystem under normal circumstances, in which data are all indexed correctly, then since those `extra data` are in the position of previous zero, those data are likely just ignored by the filesystem. Then the user is likely fine.
3. If the user is using a filesystem but experiences events like sudden power off, the recovery process (filesystem itself or fsck) might try to scan the disk to find useful information. Then those `extra blocks` will cause issues.

**Comment by c3y1huang** [comment]
Fix already included in manager and engine v1.1.2 branch.

**Comment by khushboo-rancher** [comment]
Verified with images Longhorn-master-head and v1.1.2-head `05/27/2021`

Validation - **Fail**

Ran the test `test_dr_volume_with_backup_block_deletion`, it works fine with master but observed issue https://github.com/longhorn/longhorn/issues/2642
